## Importing the Excel spreadsheets (start with ON Colleges)

Some of this has been done for ON 2022-23 in Excel as I knew how to clean in that

The excel workbooks for have multiple sheets arranged by student age, gender, status in Canada etc...I will import individual sheets and add columns of interest to one unified df with all the information of interest.

In [33]:
# open the excel file on 22-23 Stats sheet as a pandas dataframe
# read the data from the sheet

import openpyxl
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# open the excel file
wb = openpyxl.load_workbook('psi_data/cleaning_copy_excel/on_college_2012-/2012-2022 college_enrolment_headcount.xlsx')


In [34]:
wb.sheetnames

['Campus',
 'Tongue',
 'CIP',
 'Credentials',
 'Age',
 'Gender',
 'Study Level',
 'Program',
 'Status',
 'Country',
 'Combined Statistics']

In [35]:
sheet = wb['Combined Statistics']

combined_data = sheet.values

# convert to dataframe, first row is a header as it's column names
cols = next(combined_data)
df = pd.DataFrame(combined_data, columns=cols)

In [36]:
df

College Name Fiscal Year  \
0       Algonquin College   2022-2023   
1       Algonquin College   2022-2023   
2       Algonquin College   2022-2023   
3       Algonquin College   2022-2023   
4       Algonquin College   2022-2023   
..                    ...         ...   
181  St. Lawrence College   2022-2023   
182  St. Lawrence College   2022-2023   
183  St. Lawrence College   2022-2023   
184  St. Lawrence College   2022-2023   
185  St. Lawrence College   2022-2023   

                                      Status In Canada  \
0          Aboriginal ancestry (where self-identified)   
1                        Attending an off-shore school   
2                                     Canadian citizen   
3                Convention refugee (protected person)   
4    In Canada on authority of another visa (includ...   
..                                                 ...   
181  In Canada on authority of another visa (includ...   
182                                              Other   
183  Permanent resident, formerly called landed imm...   
184  Student permit/ Student visa: a permit obtaine...   
185                                            Unknown   

     Headcount Full-Time Fall 22-23  Total Full-Time Headcount Fall 22-23  \
0                                60                                   NaN   
1                                 0                                   NaN   
2                             12319                                   NaN   
3                               139                                   NaN   
4                                45                                   NaN   
..                              ...                                   ...   
181                               5                                   NaN   
182                               5                                   NaN   
183                             165                                   NaN   
184                            6939                                   NaN   
185                              35                               11243.0   

     % International 22-23  % PR 22-23  % Canadian 22-23  % Aboriginal 22-23  \
0                      NaN         NaN               NaN                 NaN   
1                      NaN         NaN               NaN                 NaN   
2                      NaN         NaN               NaN                 NaN   
3                      NaN         NaN               NaN                 NaN   
4                      NaN         NaN               NaN                 NaN   
..                     ...         ...               ...                 ...   
181                    NaN         NaN               NaN                 NaN   
182                    NaN         NaN               NaN                 NaN   
183                    NaN         NaN               NaN                 NaN   
184                    NaN         NaN               NaN                 NaN   
185               0.617184    0.014676          0.339856            0.024282   

     % Unknown 22-23  % Other 22-23  Total % Status - Audit check  None  
0                NaN            NaN                           NaN  None  
1                NaN            NaN                           NaN  None  
2                NaN            NaN                           NaN  None  
3                NaN            NaN                           NaN  None  
4                NaN            NaN                           NaN  None  
..               ...            ...                           ...   ...  
181              NaN            NaN                           NaN  None  
182              NaN            NaN                           NaN  None  
183              NaN            NaN                           NaN  None  
184              NaN            NaN                           NaN  None  
185         0.003113       0.000445                      0.999555  None  

[186 rows x 13 columns]

### Formatting issues, Pandas is reading in >1 million rows from Excel and the literal formulas, not the values. Now solved in the excel files, section can be ignored

Will clean this first, confirming below that there are 186 rows of data as there are in Excel

In [10]:
# how many rows in the datafram are completely empty?
empty_rows = df.isnull().all(axis=1).sum()
print(f"Number of totally empty rows is: {empty_rows}")
print(f"Number of rows in the dataframe is: {df.shape[0]}")
# the difference between the two is the number of rows that are not empty
print(f"Number of rows that are not empty is: {df.shape[0] - empty_rows}")

Number of totally empty rows is: 1047674
Number of rows in the dataframe is: 1047860
Number of rows that are not empty is: 186


In [11]:
# delete rows that are entirely empty
df = df.dropna(how='all')

In [13]:
df.shape

(186, 13)

In [14]:
df.head(20)

College Name Fiscal Year  \
0   Algonquin College   2022-2023   
1   Algonquin College   2022-2023   
2   Algonquin College   2022-2023   
3   Algonquin College   2022-2023   
4   Algonquin College   2022-2023   
5   Algonquin College   2022-2023   
6   Algonquin College   2022-2023   
7   Algonquin College   2022-2023   
8   Algonquin College   2022-2023   
9   Algonquin College   2022-2023   
10   Cambrian College   2022-2023   
11   Cambrian College   2022-2023   
12   Cambrian College   2022-2023   
13   Cambrian College   2022-2023   
14   Cambrian College   2022-2023   
15   Cambrian College   2022-2023   
16   Cambrian College   2022-2023   
17   Cambrian College   2022-2023   
18   Canadore College   2022-2023   
19   Canadore College   2022-2023   

                                     Status In Canada  \
0         Aboriginal ancestry (where self-identified)   
1                       Attending an off-shore school   
2                                    Canadian citizen   
3               Convention refugee (protected person)   
4   In Canada on authority of another visa (includ...   
5                                               Other   
6   Permanent resident, formerly called landed imm...   
7                                      Refugee status   
8   Student permit/ Student visa: a permit obtaine...   
9                                             Unknown   
10        Aboriginal ancestry (where self-identified)   
11                                   Canadian citizen   
12              Convention refugee (protected person)   
13                                              Other   
14  Permanent resident, formerly called landed imm...   
15                                     Refugee status   
16  Student permit/ Student visa: a permit obtaine...   
17                                            Unknown   
18        Aboriginal ancestry (where self-identified)   
19                                   Canadian citizen   

    Headcount Full-Time Fall 22-23 Total Full-Time Headcount Fall 22-23  \
0                             60.0                                 None   
1                              0.0                                 None   
2                          12319.0                                 None   
3                            139.0                                 None   
4                             45.0                                 None   
5                             23.0                                 None   
6                           1538.0                                 None   
7                             16.0                                 None   
8                           3738.0                                 None   
9                              0.0                         =SUM(D2:D11)   
10                             5.0                                 None   
11                          2747.0                                 None   
12                             5.0                                 None   
13                             0.0                                 None   
14                            93.0                                 None   
15                             5.0                                 None   
16                          4900.0                                 None   
17                             0.0                        =SUM(D12:D19)   
18                            21.0                                 None   
19                          1765.0                                 None   

   % International 22-23    % PR 22-23 % Canadian 22-23 % Aboriginal 22-23  \
0                   None          None             None               None   
1                   None          None             None               None   
2                   None          None             None               None   
3                   None          None             None               None   
4                   None          None             None            

Pandas has imported the functions where I calculated the percentage of students, I will go back and copy just the values from Excel

### Open the Credentials, Gender and Program Excel Sheets, ensure formatting is OK

In [37]:
creds_sheet = wb['Credentials']

creds_data = creds_sheet.values

# convert to dataframe, first row is a header as it's column names
cols = next(creds_data)
creds_df = pd.DataFrame(creds_data, columns=cols)

In [38]:
creds_df

College Name Fiscal Year  \
0        Algonquin College   2012-2013   
1        Algonquin College   2012-2013   
2        Algonquin College   2012-2013   
3        Algonquin College   2012-2013   
4        Algonquin College   2012-2013   
...                    ...         ...   
1159  St. Lawrence College   2022-2023   
1160  St. Lawrence College   2022-2023   
1161  St. Lawrence College   2022-2023   
1162  St. Lawrence College   2022-2023   
1163  St. Lawrence College   2022-2023   

                            Credential Type Description  \
0                                           Certificate   
1                      Degree (includes applied degree)   
2                                               Diploma   
3     General Equivalency Diploma / High School Diploma   
4                       Not Available or Not Applicable   
...                                                 ...   
1159                                        Certificate   
1160                   Degree (includes applied degree)   
1161                                            Diploma   
1162                                     Not Applicable   
1163  Other type of credential associated with a pro...   

     Headcount Full-Time Fall  
0                        2577  
1                         381  
2                       13110  
3                           0  
4                           0  
...                       ...  
1159                     2262  
1160                      755  
1161                     8227  
1162                        0  
1163                        0  

[1164 rows x 4 columns]

In [39]:
gender_sheet = wb['Gender']

gender_data = gender_sheet.values

# convert to dataframe, first row is a header as it's column names
cols = next(gender_data)
gender_df = pd.DataFrame(gender_data, columns=cols)

In [40]:
gender_df

College Name Fiscal Year       Gender Description  \
0       Algonquin College   2012-2013                   Female   
1       Algonquin College   2012-2013                     Male   
2       Algonquin College   2012-2013              Undisclosed   
3       Algonquin College   2013-2014                   Female   
4       Algonquin College   2013-2014                     Male   
..                    ...         ...                      ...   
790  St. Lawrence College   2021-2022              Undisclosed   
791  St. Lawrence College   2022-2023                   Female   
792  St. Lawrence College   2022-2023  Identify in another way   
793  St. Lawrence College   2022-2023                     Male   
794  St. Lawrence College   2022-2023              Undisclosed   

    Headcount Full-Time Fall  Total FT Enrolment 22-23  % Female 22-23  
0                       7537                       NaN             NaN  
1                       8492                       NaN             NaN  
2                         39                       NaN             NaN  
3                       8036                       NaN             NaN  
4                       8756                       NaN             NaN  
..                       ...                       ...             ...  
790                       33                       NaN             NaN  
791                     5668                       NaN             NaN  
792                       46                       NaN             NaN  
793                     5475                       NaN             NaN  
794                       55                   11244.0        0.504091  

[795 rows x 6 columns]

## Cleaning & Feature Engineering of College Gender, Credentials, Program and Status sheets

The goal is to have each record/row in the dataframe be either:
- one PSI or;
- one PSI per academic year

### Start with Credentials sheet

In [57]:
creds_df.head(10)

College Name Fiscal Year  \
0  Algonquin College   2012-2013   
1  Algonquin College   2012-2013   
2  Algonquin College   2012-2013   
3  Algonquin College   2012-2013   
4  Algonquin College   2012-2013   
5  Algonquin College   2013-2014   
6  Algonquin College   2013-2014   
7  Algonquin College   2013-2014   
8  Algonquin College   2013-2014   
9  Algonquin College   2013-2014   

                         Credential Type Description Headcount Full-Time Fall  
0                                        Certificate                     2577  
1                   Degree (includes applied degree)                      381  
2                                            Diploma                    13110  
3  General Equivalency Diploma / High School Diploma                        0  
4                    Not Available or Not Applicable                        0  
5                                        Certificate                     2856  
6                   Degree (includes applied degree)                      437  
7                                            Diploma                    13551  
8  General Equivalency Diploma / High School Diploma                        0  
9                    Not Available or Not Applicable                        0

In [44]:
creds_df.shape

(1164, 4)

Are there any credential types that are almost negligible and can be ignored?

In [43]:
# unique entries for Credential Type Description
creds_df['Credential Type Description'].value_counts()

Credential Type Description
Certificate                                             264
Diploma                                                 264
Degree (includes applied degree)                        210
Not Applicable                                          163
Not Available or Not Applicable                         132
Other type of credential associated with a program       79
General Equivalency Diploma / High School Diploma        27
Attestation and other credentials for short programs     25
Name: count, dtype: int64

I need to explore carefully - the above numbers just tell me how many times this field was recorded, nothing about headcount on those programs.

I also want to note whether these institutions **offer** certain qualifications like degrees, as well as the change in enrolment to these programs.
Structure the call step by step:

In [47]:
# sum of the number of students in each credential type
creds_df[creds_df['Credential Type Description'] == 'Certificate']

College Name Fiscal Year Credential Type Description  \
0        Algonquin College   2012-2013                 Certificate   
5        Algonquin College   2013-2014                 Certificate   
10       Algonquin College   2014-2015                 Certificate   
15       Algonquin College   2015-2016                 Certificate   
20       Algonquin College   2016-2017                 Certificate   
...                    ...         ...                         ...   
1135  St. Lawrence College   2018-2019                 Certificate   
1141  St. Lawrence College   2019-2020                 Certificate   
1147  St. Lawrence College   2020-2021                 Certificate   
1153  St. Lawrence College   2021-2022                 Certificate   
1159  St. Lawrence College   2022-2023                 Certificate   

     Headcount Full-Time Fall  
0                        2577  
5                        2856  
10                       2937  
15                       3094  
20                       3075  
...                       ...  
1135                     1067  
1141                     1150  
1147                     1022  
1153                     1431  
1159                     2262  

[264 rows x 4 columns]

In [48]:
certificates = creds_df[creds_df['Credential Type Description'] == 'Certificate']

Above: That is every school in every fiscal year that has a record of a certificate credential and the headcount.

Below: Same with Degrees, Diplomas, Not Applicable & Other

In [54]:
diplomas = creds_df[creds_df['Credential Type Description'] == 'Diploma']

In [55]:
degrees = creds_df[creds_df['Credential Type Description'] == 'Degree (includes applied degree)']

In [59]:
cred_na = creds_df[(creds_df['Credential Type Description'] == 'Not Applicable') | (creds_df['Credential Type Description'] == 'Not Available or Not Applicable')]

Above: Some of the records are 'Not Applicable', some are 'Not Available or Not Applicable' so I will combine the records to state Not Applicable below

In [76]:
creds_df.loc[
    (creds_df['Credential Type Description'] == 'Not Applicable') | 
    (creds_df['Credential Type Description'] == 'Not Available or Not Applicable'), 
    'Credential Type Description'] = 'Not Applicable'

In [78]:
cred_na = creds_df[(creds_df['Credential Type Description'] == 'Not Applicable') | (creds_df['Credential Type Description'] == 'Not Available or Not Applicable')]

In [79]:
cred_na

College Name Fiscal Year Credential Type Description  \
4        Algonquin College   2012-2013              Not Applicable   
9        Algonquin College   2013-2014              Not Applicable   
14       Algonquin College   2014-2015              Not Applicable   
19       Algonquin College   2015-2016              Not Applicable   
24       Algonquin College   2016-2017              Not Applicable   
...                    ...         ...                         ...   
1138  St. Lawrence College   2018-2019              Not Applicable   
1144  St. Lawrence College   2019-2020              Not Applicable   
1150  St. Lawrence College   2020-2021              Not Applicable   
1156  St. Lawrence College   2021-2022              Not Applicable   
1162  St. Lawrence College   2022-2023              Not Applicable   

     Headcount Full-Time Fall  
4                           0  
9                           0  
14                          0  
19                          0  
24                          0  
...                       ...  
1138                        0  
1144                        0  
1150                        0  
1156                        0  
1162                        0  

[295 rows x 4 columns]

In [80]:
# check
creds_df['Credential Type Description'].value_counts()

Credential Type Description
Not Applicable                                          295
Certificate                                             264
Diploma                                                 264
Degree (includes applied degree)                        210
Other type of credential associated with a program       79
General Equivalency Diploma / High School Diploma        27
Attestation and other credentials for short programs     25
Name: count, dtype: int64

In [86]:
cred_other = creds_df[creds_df['Credential Type Description'] == 'Other type of credential associated with a program']

In [87]:
cred_other

College Name Fiscal Year  \
104       Canadore College   2012-2013   
107       Canadore College   2013-2014   
110       Canadore College   2014-2015   
113       Canadore College   2015-2016   
116       Canadore College   2016-2017   
...                    ...         ...   
1139  St. Lawrence College   2018-2019   
1145  St. Lawrence College   2019-2020   
1151  St. Lawrence College   2020-2021   
1157  St. Lawrence College   2021-2022   
1163  St. Lawrence College   2022-2023   

                            Credential Type Description  \
104   Other type of credential associated with a pro...   
107   Other type of credential associated with a pro...   
110   Other type of credential associated with a pro...   
113   Other type of credential associated with a pro...   
116   Other type of credential associated with a pro...   
...                                                 ...   
1139  Other type of credential associated with a pro...   
1145  Other type of credential associated with a pro...   
1151  Other type of credential associated with a pro...   
1157  Other type of credential associated with a pro...   
1163  Other type of credential associated with a pro...   

     Headcount Full-Time Fall  
104                         0  
107                         0  
110                         0  
113                         0  
116                         0  
...                       ...  
1139                        0  
1145                        0  
1151                      368  
1157                      175  
1163                        0  

[79 rows x 4 columns]